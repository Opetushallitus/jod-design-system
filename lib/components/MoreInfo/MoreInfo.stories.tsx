import type { StoryObj } from '@storybook/react';
import type { TitledMeta } from '../../utils';

import { MoreInfo } from './MoreInfo';

const meta = {
  title: 'Navigation/MoreInfo',
  component: MoreInfo,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof MoreInfo>;

export default meta;

type Story = StoryObj<typeof meta>;

const links = [
  {
    to: 'ohjeet/tietoa-palvelusta',
    label: 'Tietoa palvelusta',
  },
  {
    to: 'perustiedot/tietosuojaseloste',
    label: 'Tietosuojaselosteet ja evästeet',
  },
  {
    to: 'perustiedot/datalahteet',
    label: 'Datalähteet',
  },
  {
    to: 'perustiedot/tietoa-tekoalysta',
    label: 'Tietoa tekoälyn käytöstä',
  },
  {
    to: 'perustiedot/saavutettavuusseloste',
    label: 'Saavutettavuusseloste',
  },
];

export const Default: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/a3heEFNpoI9MRcMFtbYVfO/Osaamispolku---Design-system?node-id=1048-4317',
    },
  },
  args: {
    title: 'Haluatko tietää lisää Osaamispolusta?',
    description:
      'Mietityttääkö tietosuoja tai tekoälyn hyödyntäminen palvelussa? Alta löydät kootusti yleistä tietoa palvelusta ja sen käytöstä.',
    language: 'fi',
    links: links,
    LinkComponent: ({ children, to, className }) => (
      <a href={to} className={className}>
        {children}
      </a>
    ),
  },
};
