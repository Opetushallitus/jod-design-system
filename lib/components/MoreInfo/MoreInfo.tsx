import { MdArrowForward } from 'react-icons/md';

interface InfoLinkProps {
  to: string;
  label: string;
  LinkComponent: MoreInfoProps['LinkComponent'];
}

const LinkItem = ({ to, label, LinkComponent }: InfoLinkProps) => (
  <li>
    <LinkComponent to={to} className="ds:flex ds:justify-center ds:gap-4 ds:text-button-md-mobile ds:sm:text-button-md">
      <span>{label}</span>
      <MdArrowForward size={24} />
    </LinkComponent>
  </li>
);

interface Link {
  to: string;
  label: string;
}

export interface MoreInfoProps {
  title: string;
  description: string;
  language: string;
  links: Link[];
  LinkComponent: React.ComponentType<{ children: React.ReactNode; className?: string; to: string }>;
}

export const MoreInfo = ({ title, description, language, LinkComponent, links }: MoreInfoProps) => {
  return (
    <div className="ds:flex ds:justify-start ds:text-white ds:bg-[#333] ds:py-6">
      <div className="ds:w-[1092px] ds:mx-auto ds:px-5 ds:sm:px-6">
        <div className="ds:flex ds:flex-col">
          <div className="ds:text-heading-2-mobile ds:sm:text-heading-2 ds:mb-3">{title}</div>
          <p className="ds:text-body-sm-mobile ds:sm:text-body-sm ds:mb-6">{description}</p>
          <ul className="ds:flex ds:flex-col ds:gap-3 ds:justify-start ds:items-start">
            {links.map((link) => (
              <LinkItem
                key={link.label}
                to={`/${language}/${link.to}`}
                label={link.label}
                LinkComponent={LinkComponent}
              />
            ))}
          </ul>
        </div>
      </div>
    </div>
  );
};
