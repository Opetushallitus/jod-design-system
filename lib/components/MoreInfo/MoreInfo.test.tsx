import { render, screen } from '@testing-library/react';
import React from 'react';
import { describe, expect, it } from 'vitest';
import { MoreInfo, MoreInfoProps } from './MoreInfo';

const MockLinkComponent = ({
  to,
  children,
  className,
}: {
  to: string;
  children: React.ReactNode;
  className?: string;
}) => (
  <a href={to} className={className} data-testid="mock-link">
    {children}
  </a>
);

const defaultProps: MoreInfoProps = {
  title: 'Test Title',
  description: 'Test description content.',
  language: 'en',
  links: [
    { to: 'page-1', label: 'Page 1' },
    { to: 'page-2', label: 'Page 2' },
  ],
  LinkComponent: MockLinkComponent,
};
it('matches snapshot', () => {
  const { container } = render(<MoreInfo {...defaultProps} />);
  expect(container.firstChild).toMatchSnapshot();
});

describe('MoreInfo', () => {
  it('renders the title and description', () => {
    render(<MoreInfo {...defaultProps} />);
    expect(screen.getByText('Test Title')).toBeInTheDocument();
    expect(screen.getByText('Test description content.')).toBeInTheDocument();
  });

  it('renders all links with correct labels and hrefs', () => {
    render(<MoreInfo {...defaultProps} />);
    defaultProps.links.forEach((link) => {
      const linkElement = screen.getByText(link.label).closest('a');
      expect(linkElement).toBeInTheDocument();
      expect(linkElement).toHaveAttribute('href', `/en/${link.to}`);
    });
  });

  it('renders the correct number of links', () => {
    render(<MoreInfo {...defaultProps} />);
    const links = screen.getAllByTestId('mock-link');
    expect(links).toHaveLength(defaultProps.links.length);
  });

  it('renders the arrow icon for each link', () => {
    render(<MoreInfo {...defaultProps} />);
    const links = screen.getAllByTestId('mock-link');
    links.forEach((link) => {
      expect(link.querySelector('svg')).toBeInTheDocument();
    });
  });
});
