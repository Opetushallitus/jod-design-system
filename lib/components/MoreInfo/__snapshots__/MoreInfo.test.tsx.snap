// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div
  class="ds:flex ds:justify-start ds:text-white ds:bg-[#333] ds:py-6"
>
  <div
    class="ds:w-[1092px] ds:mx-auto ds:px-5 ds:sm:px-6"
  >
    <div
      class="ds:flex ds:flex-col"
    >
      <div
        class="ds:text-heading-2-mobile ds:sm:text-heading-2 ds:mb-3"
      >
        Test Title
      </div>
      <p
        class="ds:text-body-sm-mobile ds:sm:text-body-sm ds:mb-6"
      >
        Test description content.
      </p>
      <ul
        class="ds:flex ds:flex-col ds:gap-3 ds:justify-start ds:items-start"
      >
        <li>
          <a
            class="ds:flex ds:justify-center ds:gap-4 ds:text-button-md-mobile ds:sm:text-button-md"
            data-testid="mock-link"
            href="/en/page-1"
          >
            <span>
              Page 1
            </span>
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            class="ds:flex ds:justify-center ds:gap-4 ds:text-button-md-mobile ds:sm:text-button-md"
            data-testid="mock-link"
            href="/en/page-2"
          >
            <span>
              Page 2
            </span>
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
